### PR TITLE
Add builders and engine hash to fingerprint

### LINKED ISSUE #26258: Engine dart changes require flutter clean
@GaryQian reports that the `flutter` tool doesn't consider the engine version as something that should cause a rebuild.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

The presence of builders switches the kernel generated to a mutliroot scheme. Engine version also can cause kernel incompatibilities. Include both in fingerprinter to prevent using incompatible kernel.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/26258

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
_gentle ping_

File: packages/flutter_tools/lib/src/codegen.dart [hunk 1]
```diff
@@ -148,8 +148,7 @@ class CodeGeneratingKernelCompiler implements KernelCompiler {
       fileSystemScheme: kMultiRootScheme,
       depFilePath: depFilePath,
       targetModel: targetModel,
-      // Pass an invalid file name to prevent frontend_server from initializing from dill.
-      initializeFromDill: 'none_file',
+      initializeFromDill: initializeFromDill,
```
Comment (goderbauer) [inline packages/flutter_tools/lib/src/codegen.dart]: How come we now want the dill whereas previously we specifically didnt?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/codegen.dart]: It took me a while to track down this bug - switching between no-builders/builders changes the dill contents in compatible ways due to the multiroot scheme. Adding this logic to the fingerprinter makes the invalidation automatic.

File: packages/flutter_tools/lib/src/compile.dart [hunk 2]
```diff
@@ -193,6 +200,8 @@ class KernelCompiler {
           'entryPoint': mainPath,
           'trackWidgetCreation': trackWidgetCreation.toString(),
           'linkPlatformKernelIn': linkPlatformKernelIn.toString(),
+          'engineHash': engine.version,
+          'buildersUsed': '${flutterProject != null ? await flutterProject.hasBuilders : false}',
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/compile.dart]: https://github.com/flutter/flutter/pull/29399 makes `hasBuilders` synchronous, so you don't need to `await` if you submit that PR first.

**Review by yjbanov — APPROVED** [review]
lgtm

File: packages/flutter_tools/lib/src/compile.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/compile.dart]: That one needs to wait on another FE server update, so I'll handle it before submitting there

**Pull request merged**